src/contracts/middleware/IndexRegistry.sol

```diff
@@ -207,18 +214,44 @@ contract IndexRegistry is IIndexRegistry {
         _operatorIdToIndexHistory[operatorId][quorumNumber][_operatorIdToIndexHistory[operatorId][quorumNumber].length - 1].toBlockNumber = uint32(block.number);
     }
 
-    /**
-     * @notice remove an operator from the globalOperatorList  
-     * @param indexOfOperatorToRemove index of the operator to remove
-     */ 
-    function _removeOperatorFromGlobalOperatorList(uint32 indexOfOperatorToRemove) internal {
-        uint32 globalOperatorListLastIndex = uint32(globalOperatorList.length - 1);
-        bytes32 operatorIdToSwap;
-        if(indexOfOperatorToRemove != globalOperatorListLastIndex){
-            operatorIdToSwap = globalOperatorList[globalOperatorListLastIndex];
-            globalOperatorList[indexOfOperatorToRemove] = operatorIdToSwap;
-            emit GlobalIndexUpdate(operatorIdToSwap, indexOfOperatorToRemove);
+    /// @notice Returns the total number of operators of the service for the given `quorumNumber` at the given `blockNumber`
+    function _getTotalOperatorsForQuorumAtBlockNumber(uint8 quorumNumber, uint32 blockNumber) internal view returns (uint32){
+        // if there are no entries in the total operator history, return 0
+        if (_totalOperatorsHistory[quorumNumber].length == 0) {
+            return 0;
         }
-        globalOperatorList.pop();
+        
+        // if there is only one entry in the total operator history, return it
+        if (_totalOperatorsHistory[quorumNumber].length == 1) {
+            return _totalOperatorsHistory[quorumNumber][_totalOperatorsHistory[quorumNumber].length - 1].index;
+        }
+
+        // loop backwards through the total operator history to find the total number of operators at the given block number
+        for (uint i = _totalOperatorsHistory[quorumNumber].length - 2; i >= 0; i--) {
+            OperatorIndexUpdate memory totalOperatorUpdate = _totalOperatorsHistory[quorumNumber][i];
+            if(totalOperatorUpdate.toBlockNumber <= blockNumber){
+                return totalOperatorUpdate.index;
+            }
+        }        
+        return 0;
+    }
+    
+
+    /// @notice Returns the index of the `operatorId` at the given `blockNumber` fro the given `quorumNumber`, or max uint32 if the operator is not active in the quorum
+    function _getIndexOfOperatorForQuorumAtBlockNumber(bytes32 operatorId, uint8 quorumNumber, uint32 blockNumber) internal view returns(uint32) {
+        OperatorIndexUpdate memory operatorIndexUpdate;
+        // set to max uint32 value to indicate that the operator is not part of the quorum at all, until this is updated in the loop
+        operatorIndexUpdate.toBlockNumber = type(uint32).max; 
+        // loop forward through index history to find the index of the operator at the given block number
+        // this is less efficient than looping forwards, but is simpler logic and only called in view functions that aren't mined onchain
```

Comment:
these seem to conflict. we are indeed looping forwards here, so this can't be "less efficient than looping forwards".

Reply:
copilot :grimacing: